File: camerax_android/android/src/main/kotlin/dev/hebei/camerax_android/Pigeon.kt
package dev.hebei.camerax_android

import android.annotation.SuppressLint
import androidx.camera.core.CameraSelector
import androidx.camera.core.ZoomState
import androidx.camera.view.PreviewView.ScaleType

fun CameraSelectorArgs.toObject(): CameraSelector {
    val lensFacing = lensFacingArgs.toObject()
    return CameraSelector.Builder().requireLensFacing(lensFacing).build()
}

@SuppressLint("UnsafeOptInUsageError")
fun LensFacingArgs.toObject(): Int {
    return when (this) {
        LensFacingArgs.FRONT -> CameraSelector.LENS_FACING_FRONT
        LensFacingArgs.BACK -> CameraSelector.LENS_FACING_BACK
        LensFacingArgs.EXTERNAL -> CameraSelector.LENS_FACING_EXTERNAL
    }
}

fun ZoomState.toArgs(): ZoomStateArgs {
    val minZoomRatioArgs = minZoomRatio.toDouble()
    val maxZoomRatioArgs = maxZoomRatio.toDouble()
    val linearZoomArgs = linearZoom.toDouble()
    val zoomRatioArgs = zoomRatio.toDouble()
    return ZoomStateArgs(minZoomRatioArgs, maxZoomRatioArgs, linearZoomArgs, zoomRatioArgs)
}

fun ScaleTypeArgs.toObject(): ScaleType {
    return when (this) {
        ScaleTypeArgs.FILL_CENTER -> ScaleType.FILL_CENTER
        ScaleTypeArgs.FILL_END -> ScaleType.FILL_END
        ScaleTypeArgs.FILL_START -> ScaleType.FILL_START
        ScaleTypeArgs.FIT_CENTER -> ScaleType.FIT_CENTER
        ScaleTypeArgs.FIT_END -> ScaleType.FIT_END
        ScaleTypeArgs.FIT_START -> ScaleType.FIT_START
    }
}

File: camerax_android/lib/src/pigeon.dart
import 'package:camerax_platform_interface/camerax_platform_interface.dart';

import 'pigeon.g.dart';

extension CameraSelectorX on CameraSelector {
  CameraSelectorArgs toArgs() {
    return CameraSelectorArgs(
      lensFacingArgs: lensFacing.toArgs(),
    );
  }
}

extension LensFacingX on LensFacing {
  LensFacingArgs toArgs() {
    switch (this) {
      case LensFacing.front:
        return LensFacingArgs.front;
      case LensFacing.back:
        return LensFacingArgs.back;
      case LensFacing.external:
        return LensFacingArgs.external;
    }
  }
}

extension ScaleTypeX on ScaleType {
  ScaleTypeArgs toArgs() {
    switch (this) {
      case ScaleType.fillCenter:
        return ScaleTypeArgs.fillCenter;
      case ScaleType.fillEnd:
        return ScaleTypeArgs.fillEnd;
      case ScaleType.fillStart:
        return ScaleTypeArgs.fillStart;
      case ScaleType.fitCenter:
        return ScaleTypeArgs.fitCenter;
      case ScaleType.fitEnd:
        return ScaleTypeArgs.fitEnd;
      case ScaleType.fitStart:
        return ScaleTypeArgs.fitStart;
    }
  }
}
